src/lazy-progress.ts

```diff
@@ -0,0 +1,35 @@
+import { OutputChannel, window, ProgressLocation } from "vscode";
+
+"use strict";
+
+/**
+ * A progress bar that starts only the first time `startOrContinue` is called.
+ */
+export class LazyProgress {
+  private progress;
+  startOrContinue(
+    title: string,
+    output: OutputChannel,
+    download: Promise<any>
+  ): void {
+    if (!this.progress) {
+      this.progress = window.withProgress(
+        { location: ProgressLocation.Notification, title: title },
+        p =>
+          new Promise(resolve => {
+            output.show();
+            function complete() {
+              p.report({ increment: 100 });
+              setTimeout(() => {
+                resolve();
+              }, 2000);
+            }
+            download.then(() => {
+              output.hide();
+              complete();
+            }, complete);
```

Comment:
Minor, but here you're using `.then` and in other places you're using `await`.
I would prefer to pick a style and be consistent.

My personal preference is to avoid `async/await` as they compose poorly, but I care more about consistency.

Reply:
In this specific example, I usually go with

```typescript
download.then(output.hide).then(complete, complete)
```

which makes more obvious the "finally-like" intention.

By the way https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally should be supported by modern browser engines, so you can also try that. Not sure it works on Electron, but it seems like it should.

```typescript
download.then(output.hide).finally(complete)
```

Reply:
I refactored to `download.then(complete, complete)` but `.then(output.hide)` did not work. New version is still nicer.

Also refactored `await` to use `then` instead, my head needs more time to understand async/await